test/type_construction_test.rb

```diff
@@ -6482,4 +6482,94 @@ def test_typing_record_nilable_attribute
       end
     end
   end
+
+  def test_type_case_case_when_assignment
```

Comment:
Assignments in `when` branches without `else` were `nil`-able even if the case-when is exhaustive.